In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
file_path = Path("Resources/lending_data.csv")
df_all_data = pd.read_csv(file_path)

# Review the DataFrame
print(df_all_data.info())
print()
print(f"shape =  {df_all_data.shape}")
print()
print(f"columns = {list(df_all_data.columns)}")


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2   borrower_income   77536 non-null  int64  
 3   debt_to_income    77536 non-null  float64
 4   num_of_accounts   77536 non-null  int64  
 5   derogatory_marks  77536 non-null  int64  
 6   total_debt        77536 non-null  int64  
 7   loan_status       77536 non-null  int64  
dtypes: float64(3), int64(5)
memory usage: 4.7 MB
None

shape =  (77536, 8)

columns = ['loan_size', 'interest_rate', 'borrower_income', 'debt_to_income', 'num_of_accounts', 'derogatory_marks', 'total_debt', 'loan_status']


In [3]:
df_all_data.head(5)

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features
y = df_all_data['loan_status']

# Separate the y variable, the labels
df_data = df_all_data.copy()
df_data = df_data.drop(columns='loan_status', axis=1)

# Separate the X variable, the features
X = df_data


In [5]:
# Review the y variable Series
print(len(y))
print(y[:5])

77536
0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64


In [6]:
# Review the X variable DataFrame
print(X.shape)
X.head(3)

(77536, 7)


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(random_state=1, solver='lbfgs', max_iter=100, verbose=1)
classifier


LogisticRegression(random_state=1, verbose=1)

In [9]:
# Fit the model using training data
classifier.fit(X_train, y_train)


RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =            8     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  6.93147D-01    |proj g|=  2.18976D+04


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
 This problem is unconstrained.



At iterate   50    f=  2.34245D-02    |proj g|=  4.92631D+00

At iterate  100    f=  2.31385D-02    |proj g|=  8.33453D-04

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
    8    100    126      1     0     0   8.335D-04   2.314D-02
  F =   2.3138472790614640E-002

STOP: TOTAL NO. of ITERATIONS REACHED LIMIT                 


/opt/anaconda3/envs/dev/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.3s finished


LogisticRegression(random_state=1, verbose=1)

In [10]:
# Model Score
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Training Data Score: 0.9929323153115972
Testing Data Score: 0.9924680148576145


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [11]:
# Make a prediction using the testing data
predictions = classifier.predict(X_test)
df_predictions = pd.DataFrame({"Prediction": predictions, "Actual": y_test})
df_predictions

Prediction  Actual
60914           0       0
36843           0       0
1966            0       0
70137           0       0
27237           0       0
...           ...     ...
45639           0       0
11301           0       0
51614           0       0
4598            0       0
2793            0       0

[19384 rows x 2 columns]

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [12]:
# Generate a confusion matrix for the model
c_m = confusion_matrix(y_test, predictions)
c_m
# tn, fp, fn, tp

array([[18655,   110],
       [   36,   583]])

In [13]:
df_cm = pd.DataFrame(c_m, index=["test_negative", "test_positive"], columns=["pred_negative", "pred_positive"])
df_cm

pred_negative  pred_positive
test_negative          18655            110
test_positive             36            583

In [14]:
# Print the classification report for the model
target_names = ["healthy loan (0)", "high-risk loan (1)"]
print(classification_report(y_test, predictions, target_names=target_names))

                    precision    recall  f1-score   support

  healthy loan (0)       1.00      0.99      1.00     18765
high-risk loan (1)       0.84      0.94      0.89       619

          accuracy                           0.99     19384
         macro avg       0.92      0.97      0.94     19384
      weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
Based on the precision scores, 100% of predicted healthy loans(0) were correct, and 84% of high-risk loans(1) were predicted correctly.

Based on the recall scores, out of all the healthy loans(0) in the test dataset, 99% of them were correctly predicted, and out of all the high-risk-loans(1), 94% of them were correclty predicted.

Based on the support numbers, it seems like the biggest discrepency in this model can be attributed to data with a healthy loan (0) value, but it was predicted that they would be a high-risk loan(1).
There are 110 data points that fit this criteria.  But this would lead to conservative lending practices and would be a safe outcome for the lending institution.

The other discrepency is high risk loans that were classified as healthy loans (36 of the 619 data). This is about 16% of the high-risk loans that were incorrecly classified as healthy, and would be likely to cost the institution money. It doesn't seem like too much, but if the number of high-risk loans increased in the future, or these high risk loans are for a lot of money, this could be a significant mistake with the model that can lead to a significant decrease in revenue. 

Overall, the score for this model is 99.2%, so it is pretty accurate but could be improved by using a different algorithm or adjusting the hyperparameters within this model's algorithm.

# Module 12 Report Template

## Overview of the Analysis

In this section, describe the analysis you completed for the machine learning models used in this Challenge. This might include:

* Explain the purpose of the analysis.
* Explain what financial information the data was on, and what you needed to predict.
* Provide basic information about the variables you were trying to predict (e.g., `value_counts`).
* Describe the stages of the machine learning process you went through as part of this analysis.
* Briefly touch on any methods you used (e.g., `LogisticRegression`, or any other algorithms).

## Results

Using bulleted lists, describe the accuracy scores and the precision and recall scores of all machine learning models.

* Machine Learning Model 1:
    * Description of Model 1 Accuracy, Precision, and Recall scores.

## Summary

Summarize the results of the machine learning models, and include a recommendation on the model to use, if any. For example:

* Which one seems to perform best? How do you know it performs best?
* Does performance depend on the problem we are trying to solve? (For example, is it more important to predict the `1`'s, or predict the `0`'s? )

If you do not recommend any of the models, please justify your reasoning.


# Model2
## Using StandardScalar()
using scaled data ->
compare the two models

In [15]:
# import dependecies 
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import  accuracy_score

In [16]:
# Separate the data into labels and features
y2 = df_all_data['loan_status']

# Separate the y variable, the labels
df_data2 = df_all_data.copy()
df_data2 = df_data2.drop(columns='loan_status', axis=1)

# Separate the X variable, the features
X2 = df_data2


In [17]:
# Review the y variable Series
print(len(y2))
print(y2[:5])

77536
0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64


In [18]:
# Review the X variable DataFrame
print(X2.shape)
X2.head(3)

(77536, 7)


loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100

In [19]:
# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train2, X_test2, y_train2, y_test2 = train_test_split(X2, y2, random_state=1)

### SCALE

In [20]:
# Creating StandardScaler instance
scaler = StandardScaler()
# Fitting Standard Scaller
X_scaler = scaler.fit(X_train2)
# Scaling data
X_train_scaled = X_scaler.transform(X_train2)
X_test_scaled = X_scaler.transform(X_test2)




### Model creation, fit, predict

In [21]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier2 = LogisticRegression(random_state=1, solver='lbfgs', max_iter=100, verbose=1)

# Fit the model using training data
classifier2.fit(X_train_scaled, y_train2)

# Making predictions using the testing data
predictions2 = classifier2.predict(X_test_scaled)



RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =            8     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  6.93147D-01    |proj g|=  4.67654D-01

           * * *

Tit   = total number of iterations
Tnf   = total number of function evaluations
Tnint = total number of segments explored during Cauchy searches
Skip  = number of BFGS updates skipped
Nact  = number of active bounds at final generalized Cauchy point
Projg = norm of the final projected gradient
F     = final function value

           * * *

   N    Tit     Tnf  Tnint  Skip  Nact     Projg        F
    8     29     30      1     0     0   9.378D-05   1.889D-02
  F =   1.8892111548927268E-002

CONVERGENCE: NORM_OF_PROJECTED_GRADIENT_<=_PGTOL            


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
 This problem is unconstrained.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.1s finished


### Scores and results

In [22]:
# Model Score
print(f"Training Data Score: {classifier2.score(X_train_scaled, y_train2)}")
print(f"Testing Data Score: {classifier2.score(X_test_scaled, y_test2)}")

# Generate a confusion matrix for the model
c_m2 = confusion_matrix(y_test2, predictions2) # tn, fp, fn, tp
print()
print("confusion matrix for scaled data is:")
print(c_m2)
print()

df_cm2 = pd.DataFrame(c_m2, index=["test_negative", "test_positive"], columns=["pred_negative", "pred_positive"])
df_cm2


Training Data Score: 0.9942736277342138
Testing Data Score: 0.9937061494015683

confusion matrix for scaled data is:
[[18652   113]
 [    9   610]]



pred_negative  pred_positive
test_negative          18652            113
test_positive              9            610

In [23]:
# Print the classification report for the model
target_names = ["healthy loan (0)", "high-risk loan (1)"]
print(classification_report(y_test2, predictions2, target_names=target_names))

                    precision    recall  f1-score   support

  healthy loan (0)       1.00      0.99      1.00     18765
high-risk loan (1)       0.84      0.99      0.91       619

          accuracy                           0.99     19384
         macro avg       0.92      0.99      0.95     19384
      weighted avg       0.99      0.99      0.99     19384



Based on the precision scores, 100% of predicted healthy loans(0) were correct, and 84% of high-risk loans(1) were predicted correctly.

Based on the recall scores, out of all the healthy loans(0) in the test dataset, 99% of them were correctly predicted, and out of all the high-risk-loans(1), 99% of them were correclty predicted.

This model has improved the recall for high-risk loans (1), meaning that it has improved the amount of high-risk loans identified. 

The type of prediction that would be most likely to lose the most money would be a failure to identify high-risk loans. This second scaled model has decreased the number of these predictions from 36 to 9 in the test-data and from 94% to 99% identification. 


This model has increased the accuracy from 99.29 to 99.42 for the training data, and from 99.24 to 99.37 for the test data.


---